KAFKA-4039: Fix deadlock during shutdown due to log truncation not allowed

ijuma opened the pull request.

Commit:
KAFKA-4039: delay invocation of System.exit via FatalExitException

Commit:
KAFKA-4039 change FatalExitException to Error

Commit:
KAFKA-4039 fix the default value of test mode in FatalExitError

Commit:
KAFKA-4039 mock System.exit

Commit:
KAFKA-4039 exit inside an anonymous thread, add test for deadlock

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-4039-deadlock-during-shutdown

* apache/trunk: (552 commits)
  MINOR: JavaDoc markup cleanup
  KAFKA-4679: Remove unstable markers from Connect APIs
  KAFKA-4450; Add upgrade tests for 0.10.1 and rename TRUNK to DEV_BRANCH to reduce confusion
  KAFKA-4635; Client Compatibility follow-ups
  MINOR: update JavaDocs for Kafka Streams DSL helpers
  KAFKA-4557; Handle Producer.send correctly in expiry callbacks
  MINOR: Update copyright year in the NOTICE file.
  KAFKA-4664; Update docs/protocol.html with KIP-97 information
  KAFKA-4704; Coordinator cache loading fails if groupId is reused for offset storage after group is removed
  MINOR: Replace for within for each; replace if-elseif with match
  KAFKA-4644: Improve test coverage of StreamsPartitionAssignor
  MINOR: Update KTable JavaDoc
  MINOR: Include more detail in `ConfigDef.parseType` exception message
  KAFKA-4578; Upgrade notes for 0.10.2.0
  MINOR: Streams API JavaDoc improvements
  MINOR: Add Streams system test for broker backwards compatibility
  MINOR: Close create topics policy during shutdown and more tests
  MINOR: Update JavaDoc for DSL PAPI-API
  KAFKA-4636; Per listener security settings overrides (KIP-103)
  MINOR: Change logging level for ignored maybeAddMetric from debug to trace
  ...

ijuma committed:
Try a slightly different approach to solving the shutdown deadlock

* Open the latch in `ShutdownableThread` before calling `exit`
* Introduce `Exit` classes that perform exit/halt and can be changed in tests
* Invoke `exit` from the calling thread instead of spawning a new thread
* Updated the tests
* A few clean-ups

ijuma committed:
A few style clean-ups

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-4039-deadlock-during-shutdown

* apache/trunk:
  MINOR: Logging improvements in consumer internals
  KAFKA-2955; Add a simple ">" prompt to console producer
  KAFKA-4613: Follow-up to fix JavaDocs
  KAFKA-4613: Treat null-key records the same way for joins and aggreations

ijuma commented:
Continuing the work that @maysamyabandeh started in https://github.com/apache/kafka/pull/1755

ijuma committed:
Various fixes including checkstyle

wushujames commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -89,9 +89,10 @@ class KafkaRequestHandlerPool(val brokerId: Int,
       handler.shutdown
-    for(thread <- threads)
+    for (thread <- threads)
+    for (thread <- threads)
```
Duplicate line?

ijuma replied:
Good catch, not sure what happened there. Thanks for the lightning quick review, the PR is still WIP, I should have mentioned it in the title.

wushujames commented on core/src/main/scala/kafka/server/KafkaServerStartable.scala:
```diff
@@ -35,26 +35,22 @@ class KafkaServerStartable(val serverConfig: KafkaConfig, reporters: Seq[KafkaMe
-        fatal("Fatal error during KafkaServerStable shutdown. Prepare to halt", e)
+      case _: Throwable =>
+        fatal("Halting Kafka.")
         // Calling exit() can lead to deadlock as exit() can be called multiple times. Force exit.
```
Is this comment still needed, since this PR is attempting to solve that exact problem?

ijuma replied:
Yeah, that is still needed. The issue happens if there is a failure during `KafkaServerStartable.startup` causing `KafkaServerStartable.shutdown` to be invoked, which will fail causing it to be invoked again if `System.exit` is called (due to the shutdown hook).

There are a few ways we could fix this, but I thought I'd leave it as is for now.

ijuma committed:
Add missing `FatalExitError` rethrows

ijuma committed:
Add missing tearDown in test

ijuma committed:
Fix `KafkaConfigTest` failures

ijuma committed:
Fix bug where `Exit.exit` was called instead of `System.exit`

ijuma committed:
Add integration test, fix bugs and clean-ups

ijuma commented:
I think this is ready for review, cc @junrao @hachikuji

junrao commented on core/src/main/scala/kafka/utils/Exit.scala:
```diff
@@ -0,0 +1,53 @@
+
+  def exit(statusCode: Int, message: Option[String] = None): Nothing = {
+    JExit.exit(statusCode, message.orNull)
+    throw new AssertionError("exit should not return, but it did.")
```
Should this be done in Exit.java? Ditto for halt().

ijuma replied:
I added this because it won't compile otherwise given the `Nothing` return type. I can add it to the Java implementation as well if you think it's useful.

junrao replied:
Ah, then, we don't need that in the java implementation.

junrao commented on core/src/test/scala/integration/kafka/server/ReplicaFetcherThreadFatalErrorTest.scala:
```diff
@@ -0,0 +1,145 @@
+      }
+    }))
+    createTopic(zkUtils, "topic")
+    TestUtils.waitUntilTrue(() => shutdownCompleted, "Shutdown of follower did not complete", waitTime = 120000)
```
Do we really need to wait for 120 secs? Ditto below.

ijuma replied:
Good catch, I had added that while debugging an issue and it should not be needed. Will remove.

junrao commented on core/src/test/scala/unit/kafka/utils/ShutdownableThreadTest.scala:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    thread.start()
+    TestUtils.waitUntilTrue(() => !thread.isRunning.get, "Thread did not transition to `running` state")
```
The text should read  "Thread did not transition to `not running` state"? Also, do we need both this and the above test?

ijuma replied:
Yes, it should read like you suggest indeed. Will fix. It's a good question if we need both tests. I'll remove `testShutdownWhenCalledAfterFatalExitExceptionIsThrown` because it's covered by the integration test and `testShutdownWhenCalledAfterThreadStart` kind of tests this case too depending on the thread interleavings.

junrao reviewed (commented):
@ijuma : Thanks for the thorough patch. LGTM. Just a few minor comments.

ijuma committed:
Address review feedback

ijuma commented:
Thanks for the review @junrao, I addressed your feedback. Also, if you merge the PR, please keep @maysamyabandeh as the main author (I believe the merge script will do that by default).

wushujames commented:
I read through KAFKA-4039. Does this PR (and that JIRA) only apply for when the broker itself initiates shutdown (like during error scenarios that it self-detects)? Does this same mechanism also apply if an operator initiates a controlled shutdown (by using the kafka-server-stop.sh script)?

ijuma commented:
If the shutdown is initiated via SIGTERM, then a JVM thread starts the `exit` procedure and the Kafka shutdown hook is eventually invoked which calls KafkaServer.shutdown(). In such cases, deadlocks are less likely because we avoid the two problems addressed by this PR: calling System.exit() with locks held or while blocking a thread from termination (due to closed latch or joined thread). Does that make sense?

resetius commented:
Thanks! It fixes my issue.

junrao commented:
@ijuma : Thanks the updated patch. LGTM. I will let you merge it.
